library/base.go

```diff
@@ -0,0 +1,227 @@
+	mtx         sync.Mutex
+	cgi         CgroupImpl
+}
+
+func NewRunner(a AuthZRules, cgi CgroupImpl) *RunnerImpl {
+	cwd, err := os.Getwd()
+	if err != nil {
+		panic(err)
+	}
+
+	wrapper := fmt.Sprintf("%s/../wrapper/wrapper", cwd)
+	wrapper = path.Clean(wrapper)
+
+	return &RunnerImpl{
+		running:     map[Process]*runningProcess{},
+		wrapperPath: wrapper,
+		cgi:         cgi,
+		authz:       a,
+	}
+}
+
+var _ Runner = (*RunnerImpl)(nil)
+
+func (a *AuthZRules) getControlGroup(name string) (*ControlGroup, error) {
+	for _, cg := range a.ControlGroups {
+		if cg.Name == name {
+			return &cg, nil
+		}
+	}
+	return nil, errors.New("no such control group name")
+}
+
+func (r *RunnerImpl) internalCleanup(puuid Process) {
+	log.Println("looking up process", puuid)
+	fn := fmt.Sprintf("/sys/fs/cgroup/%s", puuid)
+	if err := os.RemoveAll(fn); err != nil {
+		log.Fatalln("cannot clean up cgroup dir", fn, err)
+	}
+
+	// clean up temp dir
+	if rp, ok := r.getRunningProcess(puuid); ok {
+		log.Println("cleaning up", rp.tempdir)
+		if err := os.RemoveAll(rp.tempdir); err != nil {
+			log.Fatalln("cannot clean up temp dir", rp.tempdir, err)
+		}
+	} else {
+		log.Println("could not find process", puuid)
+	}
+}
+
+func cleanup(r *RunnerImpl, puuid Process) func() {
+	return func() {
+		r.internalCleanup(puuid)
+	}
+}
+
+func (r *RunnerImpl) addRunningProcess(puuid Process, rp *runningProcess) {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	r.running[puuid] = rp
+}
+
+func (r *RunnerImpl) getRunningProcess(puuid Process) (*runningProcess, bool) {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	rp, ok := r.running[puuid]
+	return rp, ok
+}
+
+func (r *RunnerImpl) removeRunningProcess(puuid Process) {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	delete(r.running, puuid)
+}
+
+func (r *RunnerImpl) Run(ctx context.Context, client string, cmd string, args ...string) (Process, error) {
+	cgn, ok := r.authz.ClientToCGroup[client]
+	if !ok {
+		return "", errors.New("authorization failed - no cgroup defined for this client")
+	}
+	cg, err := r.authz.getControlGroup(cgn)
+	if err != nil {
+		return "", err
+	}
+
+	puuid := Process(uuid.New().String())
+
+	dir, err := ioutil.TempDir("", string(puuid)+"_*")
+	if err != nil {
+		return "", err
+	}
+
+	os.Chdir(dir)
+
+	if err := r.cgi.MakeCgroup(cg, puuid); err != nil {
+		log.Fatalln("unable to make cgroup:", err)
+	}
+
+	cgDir := fmt.Sprintf("/sys/fs/cgroup/%s", puuid)
+
+	var buf SafeBuffer
+	ps, st, done, err := r.cgi.Run(ctx, r.wrapperPath, cg.SysProcAttr, &buf, cleanup(r, puuid), cgDir, cmd, args...)
+	if err != nil {
+		r.internalCleanup(puuid)
+		return "", err
+	}
+
+	if ps.Process != nil {
+		r.addRunningProcess(puuid, &runningProcess{cmd: ps, output: &buf, tracker: st, pid: ps.Process.Pid, tempdir: dir, done: done})
+		return puuid, nil
+	}
+	r.internalCleanup(puuid)
+	return "", errors.New("process did not launch")
+}
+
+func (r *RunnerImpl) safeGetRunningProcess(p Process) (*runningProcess, error) {
+	rp, ok := r.running[p]
```

Comment:
This a data race on `running`, it should hold the lock while reading the `*runningProcess`.

Reply:
thanks ... I broke that method out for that purpose and forgot to actually use the lock :/